imap/jmap_calendar.c

```diff
@@ -1675,77 +1671,67 @@ static int jmap_calendarevent_getblob(jmap_req_t *req,
-        write_body(HTTP_OK, req->txn, buf_base(&buf), buf_len(&buf));
-        free(content_type);
-        res = HTTP_OK;
+        buf_setcstr(&ctx->blob, icalcomponent_as_ical_string(ical));
     }
     else {
-        /* Iniitialize multipart body */
-        if (accept_mime && strcmp(accept_mime, "multipart/mixed")) {
-            res = HTTP_NOT_ACCEPTABLE;
-            goto done;
-        }
-        req->txn->resp_body.type = "multipart/mixed";
-        write_multipart_body(HTTP_OK, req->txn, NULL, 0, NULL);
+        /* Create multipart body */
+        const char *preamble =
+            "This is a message with multiple parts in MIME format.\r\n";
+        const char *epilogue = "\r\nEnd of MIME multipart body.\r\n";
+        char boundary[100];
+        struct buf *blob = &ctx->blob;
 
-        /* Set main component Content headers */
-        char *content_type = NULL;
-        const char *comp_type = caldav_comp_type_as_string(cdata->comp_type);
-        if (comp_type) {
-            content_type = strconcat("text/calendar; component=", comp_type, NULL);
-            req->txn->resp_body.type = content_type;
-        }
-        else req->txn->resp_body.type = "text/calendar";
+        snprintf(boundary, sizeof(boundary), "%s-%ld-%ld-%ld",
+                 *spool_getheader(req->txn->req_hdrs, ":authority"),
+                 (long) getpid(), (long) time(0), (long) rand());
+
+        buf_reset(&cbuf);
+        buf_printf(&cbuf, "multipart/mixed; boundary=\"%s\"", boundary);
+        ctx->content_type = buf_cstring(&cbuf);
+
+        buf_setcstr(blob, preamble);
 
         /* Write main component body */
-        buf_setcstr(&buf, icalcomponent_as_ical_string(ical));
-        req->txn->resp_body.len = buf_len(&buf);
-        write_multipart_body(0, req->txn, buf_base(&buf), buf_len(&buf), NULL);
+        buf_printf(blob, "\r\n--%s\r\n", boundary);
+        buf_appendcstr(blob, "Content-Type: text/calendar");
+        
+        const char *comp_type = caldav_comp_type_as_string(cdata->comp_type);
+        if (comp_type) buf_printf(blob, "; component=%s", comp_type);
+        buf_appendcstr(blob, "\r\n");
+
+        const char *icalstr = icalcomponent_as_ical_string(ical);
+        buf_printf(blob, "Content-Length: %zu\r\n", strlen(icalstr));
+        buf_appendcstr(blob, "\r\n");
+        buf_appendcstr(blob, icalstr);
 
         /* Write userdata parts */
-        struct calendarevent_getblob_rock rock = { req, &buf };
+        struct calendarevent_getblob_rock rock = { boundary, blob };
         annotatemore_findall(cdata->dav.mailbox, cdata->dav.imap_uid,
                              PER_USER_CAL_DATA, 0, _calendarevent_getblob_cb,
                              &rock, 0);
 
-        write_multipart_body(0, req->txn, NULL, 0, NULL);
-
-        free(content_type);
-        res = HTTP_OK;
+        /* Write close-delimiter and epilogue */
+        buf_printf(blob, "\r\n--%s--\r\n%s", boundary, epilogue);
```

Comment:
Shouldn't the epilogue be followed by the final boundary? The preamble is following the initial boundary string.